OpenSim/Tests/testIterators/testIterators.cpp

```diff
@@ -62,6 +63,110 @@ const int expectedNumJntComponents = 3;
+    }
+
+    double getTension(const SimTK::State& s) const {
+        return getComponent<PathActuator>("cableAtoB").computeActuation(s);
+    }
+
+    double getPower(const SimTK::State& s) const {
+        return getComponent<PathActuator>("cableAtoB").getPower(s);
+    }
+
+    double getHeight(const SimTK::State& s) const {
+        static const std::string hopperHeightCoord = "/Dennis/slider/yCoord";
+        return getModel().getComponent(hopperHeightCoord)
+            .getOutputValue<double>(s, "value");
+    }
+
+    double getCenterOfMassHeight(const SimTK::State& s) const {
+        SimTK::Vec3 com_position = getModel().calcMassCenterPosition(s);
+        return com_position[SimTK::YAxis];
+    }
+
+protected:
+    // Change the color of the device's path as its tension changes.
+    void extendRealizeDynamics(const SimTK::State& s) const override {
+        const auto& actuator = getComponent<PathActuator>("cableAtoB");
+        double level = fmin(1., getTension(s) / actuator.get_optimal_force());
+        actuator.getGeometryPath().setColor(s, SimTK::Vec3(0.1, level, 0.1));
+    }
+
+}; // end of Device
+    
+} // namespace OpenSim
+
+
+void testComponentListRefs() {
+    using SimTK::Vec3;
+    using SimTK::Inertia;
+
+    Model model(modelFilename);
+
+    auto device = new OpenSim::Device();
+    device->setName("device");
+
+    auto humerus = new OpenSim::Body("device_humerus", 1, Vec3(0), Inertia(0));
+    auto radius  = new OpenSim::Body("device_radius",  1, Vec3(0), Inertia(0));
+
+    auto shoulder = new OpenSim::PinJoint("device_shoulder",
+                                          model.getGround(), Vec3(0), Vec3(0),
+                                          *humerus, Vec3(0, 1, 0), Vec3(0));
+    auto elbow = new OpenSim::PinJoint("device_elbow",
+                                       *humerus, Vec3(0), Vec3(0),
+                                       *radius, Vec3(0, 1, 0), Vec3(0));
+
+    device->addComponent(shoulder);
+    device->addComponent(elbow);
+
+    model.addModelComponent(device);
+    model.finalizeFromProperties();
+
+    std::set<const Joint*> joints1{}, joints2{};
+    std::set<const Coordinate*> coords{};
+
+    std::cout << "Joints in the model: " << std::endl;
+    for(const auto& joint : model.getComponentList<Joint>()) {
+        std::cout << "    " << joint.getFullPathName() << std::endl;
+        joints1.insert(&joint);
+    }
+
+    std::cout << "Joints and Coordinates: " << std::endl;
+    for(const auto& joint : model.getComponentList<Joint>()) {
+        joints2.insert(&joint);
+        std::cout << "    Joint: " << joint.getFullPathName() << std::endl;
+        for(const auto& coord : joint.getComponentList<Coordinate>()) {
+            std::cout << "        Coord: "
+                      << coord.getFullPathName() << std::endl;
+            coords.insert(&coord);
+        }
+    }
+
+    ASSERT(joints1.size() == 4);
+    ASSERT(joints1.size() == joints2.size());
```

Comment:
I think it is important to verify that the component lists are identical, including that they produce the same order. If the order changes for any reason it means that the inner list of coordinates is affecting the outer list, which is what we are trying to check does not happen.  While `std::set` is useful to ensure uniqueness it will also order the elements, so comparing sets will not verify if list order was preserved.